packages/labs/ssr/src/lib/render-result.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ * A rendered value as an iterable of strings or Promises of string, or a
+ * nested RenderResult.
+ *
+ * This type is a synchronous Iterable so that consumers do not have to await
+ * every value according to the JS asynchronous iterator protocol, which would
+ * cause additional overhead compared to a sync iterator.
+ *
+ * Consumers should check the type of each value emitted by the iterator, and
+ * if it is a Promise await it if possible, or throw an error.
+ *
+ * The utility functions {@link collectRenderResult} and
+ * {@link collectRenderResultSync} do this for you.
+ */
+export type RenderResult = Iterable<
+  string | RenderResult | Promise<string | RenderResult>
```

Comment:
I do think this type is simpler as 
```ts
type RenderResult = Iterable<string | Promise<RenderResult>>
```
If we don't know where we'd need the `Iterable<RenderResult>` part, I'd prefer not complicating the type and the runtime checks with it. If it's just a matter of letting someone `yield someIterableFn()` rather than `yield *someIterableFn()`, seems better for the type to require just always `yield *` than require the extra check in the hot loop.

If there is some pattern this unlocks that I'm not considering, I'd agree we should make the change now. Not clear what that is.

Reply:
Reading up on generator / async generator perf it does seem that in async generators at least, `yield *` is very costly. It seems possible, even likely, that `yield *` is more clostly than yielding a nestest iterable for sync iterators too, though I definitely have not benchmarked that (and it would have to be compared against the duck-type check for `[Symbol.iterator]`)